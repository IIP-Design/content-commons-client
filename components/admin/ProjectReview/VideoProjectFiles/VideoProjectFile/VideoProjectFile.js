import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { Embed, Grid } from 'semantic-ui-react';
import moment from 'moment'; // already being used so import here

import {
  formatBytes,
  getStreamData,
  getYouTubeId,
  getVimeoId,
  secondsToHMS,
} from 'lib/utils';

const VideoProjectFile = ( { file, thumbnail } ) => {
  const [thumbnailProps, setThumbnailProps] = useState( { signedUrl: '', alt: '' } );

  useEffect( () => {
    if ( thumbnail && thumbnail.image ) {
      const { signedUrl } = thumbnail.image;
      const alt = file && file.language ? `a thumbnail image for this file in ${file.language.displayName}` : '';

      setThumbnailProps( { signedUrl, alt } );
    }
  }, [file, thumbnail] );

  if ( !file || Object.keys( file ).length === 0 ) return null;

  const {
    createdAt,
    dimensions: { height, width },
    duration,
    filename,
    filesize,
    quality,
    stream,
    use: { name: videoType },
    videoBurnedInStatus,
  } = file;

  const youTubeUrl = getStreamData( stream, 'youtube', 'url' );
  const vimeoUrl = getStreamData( stream, 'vimeo', 'url' );

  return (
    <Grid.Row className="project_unit_files">
      <Grid.Column mobile={ 16 } tablet={ 8 } computer={ 8 } className="file_embed_thumbnail">
        { youTubeUrl && (
          <Embed
            id={ getYouTubeId( youTubeUrl ) }
            placeholder={ thumbnailProps.signedUrl }
            source="youtube"
          />
        ) }
        { !youTubeUrl && vimeoUrl && (
          <Embed
            id={ getVimeoId( vimeoUrl ) }
            placeholder={ thumbnailProps.signedUrl }
            source="vimeo"
          />
        ) }
        { !youTubeUrl && !vimeoUrl && thumbnailProps.signedUrl && (
          <figure className="thumbnail overlay">
            <img
              className="thumbnail-image"
              src={ thumbnailProps.signedUrl }
              alt={ thumbnailProps.alt }
            />
          </figure>
        ) }
      </Grid.Column>

      <Grid.Column mobile={ 16 } tablet={ 8 } computer={ 8 } className="file_meta">
        <p>
          <b className="label">File Name:</b>
          { ` ${filename}` }
        </p>
        <p>
          <b className="label">Filesize:</b>
          { ` ${formatBytes( filesize )}` }
        </p>
        <p>
          <b className="label">Dimensions:</b>
          { ` ${width} x ${height}` }
        </p>
        <p>
          <b className="label">Uploaded:</b>
          { ' ' }
          <time dateTime={ createdAt }>{ `${moment( createdAt ).format( 'LL' )}` }</time>
        </p>
        <p>
          <b className="label">Duration:</b>
          { ` ${secondsToHMS( duration )}` }
        </p>
        <p>
          <b className="label">On-Screen Text:</b>
          { ` ${videoBurnedInStatus === 'SUBTITLED' ? 'Yes' : 'No'}` }
        </p>
        <p>
          <b className="label">Video Type:</b>
          { ` ${videoType}` }
        </p>
        <p>
          <b className="label">Quality:</b>
          { ` ${quality}` }
        </p>
        <p>
          <b className="label">
            { ` ${youTubeUrl ? 'YouTube' : 'Vimeo'} URL:` }
          </b>
          { ` ${youTubeUrl || vimeoUrl}` }
        </p>
      </Grid.Column>
    </Grid.Row>
  );
};

VideoProjectFile.propTypes = {
  file: PropTypes.object,
  thumbnail: PropTypes.oneOfType( [
    PropTypes.object,
    PropTypes.bool,
  ] ),
};

VideoProjectFile.defaultProps = {
  file: {},
  thumbnail: {},
};

export default VideoProjectFile;
