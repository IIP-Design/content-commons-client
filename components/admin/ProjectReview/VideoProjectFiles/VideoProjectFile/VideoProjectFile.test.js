import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';

import VideoProjectFile from './VideoProjectFile';

jest.mock( 'lib/utils', () => ( {
  formatBytes: jest.fn( () => '631.9 MB' ),
  formatDateTime: jest.fn( () => 'May 5, 2019' ),
  getStreamData: jest.fn( ( stream, site = 'youtube', field = 'url' ) => {
    const uri = stream.find( s => s.site.toLowerCase() === site );

    if ( uri && Object.keys( uri ).length > 0 ) {
      return uri[field];
    }

    return null;
  } ),
  getVimeoId: jest.fn( () => '340239507' ),
  getYouTubeId: jest.fn( () => '1evw4fRu3bo' ),
  secondsToHMS: jest.fn( () => '9:16' ),
  contentRegExp: jest.fn( () => false ),
} ) );

const id = '123';
const props = {
  file: {
    __typename: 'VideoFile',
    id: 'f19',
    createdAt: '2019-03-05T20:18:54.032Z',
    updatedAt: '2019-06-19T17:38:37.502Z',
    duration: 556,
    filename: 'video-file-1.mp4',
    filesize: 662595174,
    filetype: 'video/mp4',
    quality: 'WEB',
    url: `2019/06/${id}/video-file-1.mp4`,
    videoBurnedInStatus: 'CLEAN',
    dimensions: {
      __typename: 'Dimensions',
      id: 'd21',
      height: '1080',
      width: '1920',
    },
    language: {
      __typename: 'Language',
      id: 'en38',
      displayName: 'English',
      languageCode: 'en',
      locale: 'en-us',
      nativeName: 'English',
      textDirection: 'LTR',
    },
    use: {
      __typename: 'VideoUse',
      id: 'us31',
      name: 'Full Video',
    },
    stream: [
      {
        __typename: 'VideoStream',
        id: 'st93',
        site: 'YouTube',
        url: 'https://www.youtube.com/watch?v=1evw4fRu3bo',
      },
      {
        __typename: 'VideoStream',
        id: 'st35',
        site: 'Vimeo',
        url: 'https://vimeo.com/340239507',
      },
    ],
  },
  thumbnail: {
    __typename: 'Thumbnail',
    id: 'th11',
    size: 'FULL',
    image: {
      __typename: 'ImageFile',
      id: 'im28',
      createdAt: '2019-03-06T13:11:48.043Z',
      updatedAt: '2019-06-18T14:58:10.024Z',
      filename: 'image-1.jpg',
      filesize: 28371,
      filetype: 'image/jpeg',
      alt: null,
      url: `https://s3-bucket-url.s3.amazonaws.com/2019/06/${id}/image-1.jpg`,
      signedUrl: `https://s3-bucket-url.s3.amazonaws.com/2019/06/${id}/image-1.jpg?AWSAccessKeyId=SOMEAWSACCESSKEY&Expires=1572028336&Signature=SOMESIGNATURE`,
      language: {
        __typename: 'Language',
        id: 'en38',
        displayName: 'English',
        languageCode: 'en',
        locale: 'en-us',
        nativeName: 'English',
        textDirection: 'LTR',
      },
    },
  },
};

const Component = <VideoProjectFile { ...props } />;

const youTubeUrl = 'https://www.youtube.com/watch?v=1evw4fRu3bo';
const vimeoUrl = 'https://vimeo.com/340239507';

describe( '<VideoProjectFile />', () => {
  it( 'renders without crashing', () => {
    const wrapper = mount( Component );

    expect( wrapper.exists() ).toEqual( true );
    expect( toJSON( wrapper ) ).toMatchSnapshot();
  } );

  it( 'renders null if file prop is {}', () => {
    const wrapper = mount( Component );

    wrapper.setProps( { file: {} } );
    expect( wrapper.html() ).toEqual( null );
  } );

  it( 'renders null if file prop is null', () => {
    const wrapper = mount( Component );

    wrapper.setProps( { file: null } );
    expect( wrapper.html() ).toEqual( null );
  } );

  it.skip( 'renders an embedded YouTube video', () => {
    const wrapper = mount( Component );
    const embed = wrapper.find( 'Embed' );
    const thumbnail = wrapper.find( 'figure.thumbnail' );
    const { signedUrl } = props.thumbnail.image;

    console.log( wrapper.debug() );

    expect( embed.exists() ).toEqual( true );
    expect( embed.length ).toEqual( 1 );
    expect( embed.prop( 'id' ) ).toEqual( '1evw4fRu3bo' );
    expect( embed.prop( 'source' ) ).toEqual( 'youtube' );
    expect( embed.prop( 'placeholder' ) ).toEqual( signedUrl );
    expect( wrapper.contains( youTubeUrl ) ).toEqual( true );
    expect( wrapper.contains( vimeoUrl ) ).toEqual( false );
    expect( wrapper.contains( thumbnail ) ).toEqual( false );
  } );

  it.skip( 'renders an embedded Vimeo video if there is no YouTube url', async () => {
    const vimeoProps = {
      ...props,
      file: {
        ...props.file,
        stream: [
          {
            __typename: 'VideoStream',
            id: 'st35',
            site: 'Vimeo',
            url: 'https://vimeo.com/340239507',
          },
        ],
      },
    };
    const wrapper = mount( <VideoProjectFile { ...vimeoProps } /> );
    const embed = wrapper.find( 'Embed' );
    const { signedUrl } = vimeoProps.thumbnail.image;
    const thumbnail = wrapper.find( 'figure.thumbnail' );

    expect( embed.exists() ).toEqual( true );
    expect( embed.length ).toEqual( 1 );
    expect( embed.prop( 'id' ) ).toEqual( '340239507' );
    expect( embed.prop( 'source' ) ).toEqual( 'vimeo' );
    expect( embed.prop( 'placeholder' ) ).toEqual( signedUrl );
    expect( wrapper.contains( youTubeUrl ) ).toEqual( false );
    expect( wrapper.contains( vimeoUrl ) ).toEqual( true );
    expect( wrapper.contains( thumbnail ) ).toEqual( false );
  } );

  it( 'renders a unit thumbnail if there are no YouTube or Vimeo urls', () => {
    const noStreamsProps = {
      ...props,
      file: {
        ...props.file,
        stream: [],
      },
    };
    const wrapper = mount( <VideoProjectFile { ...noStreamsProps } /> );
    const embed = wrapper.find( 'Embed' );
    const thumbnail = wrapper.find( 'figure.thumbnail' );
    const img = thumbnail.find( 'img.thumbnail-image' );
    const { signedUrl } = noStreamsProps.thumbnail.image;
    const alt = `a thumbnail image for this file in ${noStreamsProps.file.language.displayName}`;

    expect( embed.exists() ).toEqual( false );
    expect( thumbnail.exists() ).toEqual( true );
    expect( thumbnail.length ).toEqual( 1 );
    expect( img.prop( 'src' ) ).toEqual( signedUrl );
    expect( img.prop( 'alt' ) ).toEqual( alt );
    expect( wrapper.contains( youTubeUrl ) ).toEqual( false );
    expect( wrapper.contains( vimeoUrl ) ).toEqual( false );
  } );

  it( 'does not render an embedded video or thumbnail if there are no streams or thumbnail', () => {
    const noStreamsProps = {
      ...props,
      file: {
        ...props.file,
        stream: [],
      },
      thumbnail: {
        ...props.thumbnail,
        image: {
          ...props.thumbnail.image,
          signedUrl: '',
        },
      },
    };
    const wrapper = mount( <VideoProjectFile { ...noStreamsProps } /> );

    wrapper.setProps( { thumbnail: {} } );
    const embed = wrapper.find( 'Embed' );
    const thumbnail = wrapper.find( 'figure.thumbnail' );
    const img = thumbnail.find( 'img.thumbnail-image' );

    expect( embed.exists() ).toEqual( false );
    expect( thumbnail.exists() ).toEqual( false );
    expect( img.exists() ).toEqual( false );
    expect( wrapper.contains( youTubeUrl ) ).toEqual( false );
    expect( wrapper.contains( vimeoUrl ) ).toEqual( false );
  } );
} );
