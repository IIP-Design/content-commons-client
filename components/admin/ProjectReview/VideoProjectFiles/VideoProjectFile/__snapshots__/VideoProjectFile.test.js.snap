// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VideoProjectFile /> renders without crashing 1`] = `
<VideoProjectFile
  file={
    Object {
      "__typename": "VideoFile",
      "createdAt": "2019-03-05T20:18:54.032Z",
      "dimensions": Object {
        "__typename": "Dimensions",
        "height": "1080",
        "id": "d21",
        "width": "1920",
      },
      "duration": 556,
      "filename": "video-file-1.mp4",
      "filesize": 662595174,
      "filetype": "video/mp4",
      "id": "f19",
      "language": Object {
        "__typename": "Language",
        "displayName": "English",
        "id": "en38",
        "languageCode": "en",
        "locale": "en-us",
        "nativeName": "English",
        "textDirection": "LTR",
      },
      "quality": "WEB",
      "stream": Array [
        Object {
          "__typename": "VideoStream",
          "id": "st93",
          "site": "YouTube",
          "url": "https://www.youtube.com/watch?v=1evw4fRu3bo",
        },
        Object {
          "__typename": "VideoStream",
          "id": "st35",
          "site": "Vimeo",
          "url": "https://vimeo.com/340239507",
        },
      ],
      "updatedAt": "2019-06-19T17:38:37.502Z",
      "url": "2019/06/123/video-file-1.mp4",
      "use": Object {
        "__typename": "VideoUse",
        "id": "us31",
        "name": "Full Video",
      },
      "videoBurnedInStatus": "CLEAN",
    }
  }
  thumbnail={
    Object {
      "__typename": "Thumbnail",
      "id": "th11",
      "image": Object {
        "__typename": "ImageFile",
        "alt": null,
        "createdAt": "2019-03-06T13:11:48.043Z",
        "filename": "image-1.jpg",
        "filesize": 28371,
        "filetype": "image/jpeg",
        "id": "im28",
        "language": Object {
          "__typename": "Language",
          "displayName": "English",
          "id": "en38",
          "languageCode": "en",
          "locale": "en-us",
          "nativeName": "English",
          "textDirection": "LTR",
        },
        "signedUrl": "https://s3-bucket-url.s3.amazonaws.com/2019/06/123/image-1.jpg?AWSAccessKeyId=SOMEAWSACCESSKEY&Expires=1572028336&Signature=SOMESIGNATURE",
        "updatedAt": "2019-06-18T14:58:10.024Z",
        "url": "https://s3-bucket-url.s3.amazonaws.com/2019/06/123/image-1.jpg",
      },
      "size": "FULL",
    }
  }
>
  <GridRow
    className="project_unit_files"
  >
    <div
      className="row project_unit_files"
    >
      <GridColumn
        className="file_embed_thumbnail"
        computer={8}
        mobile={16}
        tablet={8}
      >
        <div
          className="eight wide computer sixteen wide mobile eight wide tablet column file_embed_thumbnail"
        >
          <Embed
            id="1evw4fRu3bo"
            placeholder="https://s3-bucket-url.s3.amazonaws.com/2019/06/123/image-1.jpg?AWSAccessKeyId=SOMEAWSACCESSKEY&Expires=1572028336&Signature=SOMESIGNATURE"
            source="youtube"
          >
            <div
              className="ui embed"
              onClick={[Function]}
            >
              <Icon
                as="i"
                name="video play"
              >
                <i
                  aria-hidden="true"
                  className="video play icon"
                  onClick={[Function]}
                />
              </Icon>
              <img
                className="placeholder"
                src="https://s3-bucket-url.s3.amazonaws.com/2019/06/123/image-1.jpg?AWSAccessKeyId=SOMEAWSACCESSKEY&Expires=1572028336&Signature=SOMESIGNATURE"
              />
            </div>
          </Embed>
        </div>
      </GridColumn>
      <GridColumn
        className="file_meta"
        computer={8}
        mobile={16}
        tablet={8}
      >
        <div
          className="eight wide computer sixteen wide mobile eight wide tablet column file_meta"
        >
          <p>
            <b
              className="label"
            >
              File Name:
            </b>
             video-file-1.mp4
          </p>
          <p>
            <b
              className="label"
            >
              Filesize:
            </b>
             631.9 MB
          </p>
          <p>
            <b
              className="label"
            >
              Dimensions:
            </b>
             1920 x 1080
          </p>
          <p>
            <b
              className="label"
            >
              Uploaded:
            </b>
             
            <time
              dateTime="2019-03-05T20:18:54.032Z"
            >
              March 5, 2019
            </time>
          </p>
          <p>
            <b
              className="label"
            >
              Duration:
            </b>
             9:16
          </p>
          <p>
            <b
              className="label"
            >
              On-Screen Text:
            </b>
             No
          </p>
          <p>
            <b
              className="label"
            >
              Video Type:
            </b>
             Full Video
          </p>
          <p>
            <b
              className="label"
            >
              Quality:
            </b>
             WEB
          </p>
          <p>
            <b
              className="label"
            >
               YouTube URL:
            </b>
             https://www.youtube.com/watch?v=1evw4fRu3bo
          </p>
        </div>
      </GridColumn>
    </div>
  </GridRow>
</VideoProjectFile>
`;
